File: 105.construct-binary-tree-from-preorder-and-inorder-traversal.js
/*
 * @lc app=leetcode id=105 lang=javascript
 *
 * [105] Construct Binary Tree from Preorder and Inorder Traversal
 */

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function (preorder, inorder) {
  const helper = (p, iLeft, iRight) => {
    if (p >= preorder.length || iLeft > iRight) {
      return null;
    }
    const i = lookup[preorder[p]];
    const root = new TreeNode(inorder[i]);
    root.left = helper(p + 1, iLeft, i - 1);
    root.right = helper(p + i - iLeft + 1, i + 1, iRight);
    return root;
  };
  const lookup = {};
  for (let i = 0; i < inorder.length; ++i) {
    lookup[inorder[i]] = i;
  }
  return helper(0, 0, inorder.length - 1);
};
// @lc code=end

File: 105.construct-binary-tree-from-preorder-and-inorder-traversal.py
#
# @lc app=leetcode id=105 lang=python3
#
# [105] Construct Binary Tree from Preorder and Inorder Traversal
#

# Time:  O(n)
# Space: O(n)

from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        def helper(start, end):
            if start > end:
                return None
            node_val = next(preorder_iter)
            node = TreeNode(node_val)
            idx = inorder_lookup[node_val]
            node.left = helper(start, idx - 1)
            node.right = helper(idx + 1, end)
            return node

        preorder_iter = iter(preorder)
        inorder_lookup = {n: i for i, n in enumerate(inorder)}
        return helper(0, len(inorder) - 1)


# @lc code=end
